File: src/test/java/de/lukaseichler/recurrentneuralnetwork/structure/LayerTest.java
package de.lukaseichler.recurrentneuralnetwork.structure;

import java.util.Arrays;
import org.testng.annotations.Test;
import com.google.common.collect.Lists;
import de.lukaseichler.recurrentneuralnetwork.LogActivation;
import de.lukaseichler.recurrentneuralnetwork.NeuralNetworkTest;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author leichler
 */
public class LayerTest extends NeuralNetworkTest {

    @Test
    public void layerWithoutNodes() throws Exception {
        Layer layer = new Layer();
        assertThat(layer.getNeuronCount()).isEqualTo(0);
    }

    @Test
    public void simpleInput() throws Exception {
        Layer layer = new Layer(1);
        assertThat(layer.calculate(Lists.newArrayList(2.0, 3.0))).containsOnly(applyDefaultActivationFunction(5.0));
    }

    @Test
    public void multipleNodes() throws Exception {
        Layer layer = new Layer(2);
        assertThat(layer.calculate(Lists.newArrayList(2.0, 3.0))).containsOnly(applyDefaultActivationFunction(5.0), applyDefaultActivationFunction(5.0));
    }

    @Test
    public void multipleNodesWithActivationFunction() throws Exception {
        Layer layer = new Layer(2, new LogActivation());
        double value = new LogActivation().apply(5);
        assertThat(layer.calculate(Lists.newArrayList(2.0, 3.0))).containsOnly(value, value);
    }

    @Test
    public void trainLayerWithPreviousWeights() throws Exception {
        Layer layer = new Layer(2);
        assertThat(layer.train(0.3, Arrays.asList(1.3, 1.2), Arrays.asList(1.0, 2.0))).hasSize(4);
    }

    @Test
    public void trainReturnsAllUpdatedWeights() throws Exception {
        Layer layer = new Layer(2);
        assertThat(layer.train(0.3, Arrays.asList(1.0, 1.0))).hasSize(4);
    }

    @Test(expectedExceptions = NullPointerException.class)
    public void trainWithNullPreviousWeights() throws Exception {
        Layer layer = new Layer(2);
        assertThat(layer.train(0.3, null, Arrays.asList(1.0, 1.0))).hasSize(4);

    }

    @Test
    public void trainWithoutPreviousWeights() throws Exception {
        Layer layer = new Layer(2);
        assertThat(layer.train(0.3, Arrays.asList(1.0, 1.0))).hasSize(4);
    }

    @Test(expectedExceptions = NullPointerException.class)
    public void trainNullInput() throws Exception {
        Layer layer = new Layer(2);
        layer.train(0.3, null, null);
    }
}

File: src/test/scala/de/lukaseichler/recurrentneuralnetwork/structure/LayerTest.scala
package de.lukaseichler.recurrentneuralnetwork.structure

import de.lukaseichler.recurrentneuralnetwork.NeuralNetworkTest
import de.lukaseichler.recurrentneuralnetwork.activation.LogActivation
import org.assertj.core.api.Assertions._
import org.scalatest.testng.TestNGSuite
import org.scalatest.{Ignore, Matchers}
import org.testng.annotations.Test

/**
 * @author Lukas Eichler
 */
class LayerTest extends TestNGSuite with Matchers with NeuralNetworkTest {

  @Test
  @throws(classOf[Exception])
  def layerWithoutNodes {
    val layer: Layer = new Layer
    assertThat(layer.getNeuronCount).isEqualTo(0)
  }

  @Test
  @throws(classOf[Exception])
  def simpleInput {
    val layer: Layer = new Layer(1)
    layer.calculate(List(2.0, 3.0)) should contain only (applyDefaultActivationFunction(5.0))
  }

  @Test
  @throws(classOf[Exception])
  def instantiationWithoutParameterLogActivation {
    val layer: Layer = new Layer
    layer.getActivationFunction.isInstanceOf[LogActivation]
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  def nullActivationFunction {
    val layer: Layer = new Layer(0, null)
  }

  @Test
  @throws(classOf[Exception])
  def multipleNodes {
    val layer: Layer = new Layer(2)
    layer.calculate(List(2.0, 3.0)) should contain only (applyDefaultActivationFunction(5.0))
  }

  @Test
  @throws(classOf[Exception])
  def multipleNodesWithActivationFunction {
    val layer: Layer = new Layer(2, new LogActivation)
    val value: Double = new LogActivation().apply(5)
    layer.calculate(List(2.0, 3.0)) should contain only (value)
  }

  @Test
  @throws(classOf[Exception])
  def trainLayerWithPreviousWeights {
    val layer: Layer = new Layer(2)
    layer.train(0.3, List(1.3, 1.3), List(1.0, 2.0)) should have size 4
  }

  @Test
  @throws(classOf[Exception])
  def trainReturnsAllUpdatedWeights {
    val layer: Layer = new Layer(2)
    layer.train(0.3, List(1.0, 1.0)) should have size 4
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  @Ignore
  def trainWithNullPreviousWeights {
    val layer: Layer = new Layer(2)
    layer.train(0.3, null, List(1.0, 1.0)) should have size 4
  }

  @Test
  @throws(classOf[Exception])
  def trainWithoutPreviousWeights {
    val layer: Layer = new Layer(2)
    layer.train(0.3, List(1.0, 1.0)) should have size 4
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  def trainNullInput {
    val layer: Layer = new Layer(2)
    layer.train(0.3, null, null)
  }
}
